packages/breakdown-trees/src/Views/visibility/SpaceLabelDecorator.ts

```diff
@@ -0,0 +1,164 @@
+import { Point3d, Range3d } from "@bentley/geometry-core";
+
+export interface SpaceDecorationCreationProps {
+  userLabel: string;
+  centerX: number;
+  centerY: number;
+  maxTextWidth: number;
+  spaceId: number;
+}
+
+export class SpaceLabelDecorator implements Decorator {
+  private _roomLabelDecoration: SpaceLabelDecoration[] = [];
+
+  constructor() {
+    IModelApp.viewManager.addDecorator(this);
+  }
+
+  public addDecoration(props: SpaceLabelDecorationProps) {
+    const decoration = new SpaceLabelDecoration(props);
+    this._roomLabelDecoration.push(decoration);
+  }
+
+  public clearDecorations() {
+    this._roomLabelDecoration = [];
+  }
+
+  public hideDecorations() {
+    this._roomLabelDecoration.forEach((decoration) => {
+      decoration.hide();
+    });
+  }
+
+  public showDecorations() {
+    this._roomLabelDecoration.forEach((decoration) => {
+      decoration.show();
+    });
+  }
+
+  public async addDecorationsForSpaces(imodel: IModelConnection, roomIds: any, onClickHandler: DecorationClickedHandler) {
+    const roomProps: any[] = await imodel.elements.getProps(roomIds[0]);
+    const origin = roomProps[0].placement.origin;
+    let minHeight = roomProps[0].placement.bbox.low[2] + origin[2];
+
+    const allRoomsProps: any[] = await imodel.elements.getProps(roomIds);
+
+    // Create the room labels
+    const spaceDecorations: any[] = [];
+
+    const massPropertiesPromises: Array<Promise<MassPropertiesResponseProps>> = [];
+    for (const roomId of roomIds) {
+      const massProps = imodel.getMassProperties({
+        operation: MassPropertiesOperation.AccumulateAreas,
+        candidates: [roomId],
+      });
+      massPropertiesPromises.push(massProps);
+    }
+
+    const massProperties = await Promise.all(massPropertiesPromises);
+
+    for (let i = 0; i < allRoomsProps.length; i++) {
+      assert(allRoomsProps.length === roomIds.length && roomIds.length === massProperties.length, "Incorrect data length");
+      if (allRoomsProps.length !== roomIds.length || roomIds.length !== massProperties.length) {
+        return;
+      }
+
+      const room = allRoomsProps[i];
+      const roomId = roomIds[i];
+      const massProps = massProperties[i];
+
+      // Compute the min and max z values of the rooms in the storey.
+      // We're using the minHeight property to assign the z value of the decorations, so that they appear on the floor.
+      const roomOrigin = room.placement.origin;
+      const roomMinZ = roomOrigin[2] + room.placement.bbox.low[2];
+      minHeight = Math.min(minHeight, roomMinZ);
+
+      const decorationData = this.createDecorationData(room, roomId, roomOrigin, massProps);
+      spaceDecorations.push(decorationData);
+
+    }
+
+    this.addSpaceDecorations(spaceDecorations, minHeight, onClickHandler);
+  }
+
+  private createDecorationData(room: any, roomId: any, roomOrigin: number[], massProps: MassPropertiesResponseProps): SpaceDecorationCreationProps | undefined {
```

Comment:
most of the comments i'v left on `any` have been essentially just recommendations, however using roomId as typed `any` seems especially wrong, i do get how this code is likely years old, but this may be a good opportunity to modernize

Reply:
Fixed the code to avoid any.